mono/mini/mini-llvm.c

```diff
@@ -3968,6 +3992,45 @@ process_call (EmitContext *ctx, MonoBasicBlock *bb, LLVMBuilderRef *builder_ref,
 	 */
 	lcall = emit_call (ctx, bb, &builder, callee, args, LLVMCountParamTypes (llvm_sig));
 
+	if (mono_aot_can_specialize (call->method)) {
+		LLVMValueRef lmethod = callee;
+
+		int num_params = LLVMCountParams (lmethod);
+		GArray *call_site_union = (GArray *) g_hash_table_lookup (ctx->module->method_to_call_info, call->method);
+
+		if (!call_site_union) {
+			call_site_union = g_array_sized_new (FALSE, TRUE, sizeof (gboolean), num_params);
+			int zero = 0;
+			for (int i = 0; i < num_params; i++)
+				g_array_insert_val (call_site_union, i, zero);
+		}
+
+		for (int i = 0; i < num_params; i++) {
+			LLVMValueRef argument = args [i];
+
+			// Add to list of tracked callers/callees
+			GList *usages = (GList *) g_hash_table_lookup (ctx->module->lvalue_to_lcalls, argument);
+			usages = g_list_prepend (usages, lcall);
+			g_hash_table_insert (ctx->module->lvalue_to_lcalls, argument, usages);
+
+			if (g_hash_table_lookup (ctx->module->module_nonnull, argument)) {
+				g_assert (i < LLVMGetNumArgOperands (lcall));
```

Comment:
Do we support variadic functions?   how does llvm represent variadic functions?   wondering if this assert would ever be false for varargs.

Reply:
We're only propagating through direct calls, which by nature have to have identical types for the call sites and the functions. 

Documentation about that added here:
https://github.com/mono/mono/pull/13697/commits/ada64db2dda63866dd92e8f6f9d77f83e42c0d7b